subprojects/core/src/integTest/groovy/org/gradle/api/NamedDomainObjectContainerIntegrationTest.groovy

```diff
@@ -97,4 +97,257 @@ class NamedDomainObjectContainerIntegrationTest extends AbstractDomainObjectCont
         expect:
         succeeds "help"
     }
+
+    def "name based filtering"() {
+        buildFile << """
+            testContainer.named({ it.endsWith("foo") }).all { element ->
+                assert element.name in ['foo', 'barfoo']
+            }
+
+            testContainer.register("foo")
+            testContainer.register("bar")
+            testContainer.register("foobar")
+            testContainer.register("barfoo")
+        """
+        expect:
+        succeeds "help"
+    }
+
+    def "custom collection not implementing named method triggers useful error message"() {
+        buildFile << """
+            objects.newInstance(MyNamedDomainObjectCollection).named { !it.isEmpty() }
```

Comment:
💅 I would make that a unit test with the custom type part of the test source set. This would make it easier on us when changing the `NamedDomainObjectCollection` type by making test compilation fail early, find usages work in the IDE instead of having to wait for that integration test to run.

Reply:
Why are we testing custom types?  `NamedDomainObjectCollection` isn't a type that should be implemented by plugins and if they have been, it's unsupported.

Reply:
Smoke tests showed that KGP does have custom `NamedDomainObjectCollection` types. The intent here is to make sure we emit a reasonable error.